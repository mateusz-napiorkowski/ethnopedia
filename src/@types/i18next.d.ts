import global from "../translations/en/global.json"
import { resources } from "../i18n"

declare module "i18next" {
}
export {}
