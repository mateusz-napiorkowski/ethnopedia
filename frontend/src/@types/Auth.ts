export type FormValues = {
    username: string
    firstName: string,
    password: string
}

export type LoginValues = {
    username: string
    password: string
}