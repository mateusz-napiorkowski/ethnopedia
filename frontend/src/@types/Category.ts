export interface Category {
    name: string;
    values?: string[];
    subcategories?: Category[];
    isSelectable?: boolean;
}