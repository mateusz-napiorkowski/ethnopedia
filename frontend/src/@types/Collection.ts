export type Collection = {
    id?: string
    description: string
    name: string
    artworksCount?: number
    categoriesCount?: number
}
