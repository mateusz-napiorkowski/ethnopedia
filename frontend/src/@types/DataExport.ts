export enum ExportExtent {
    all = "all",
    selected = "selected",
    searchResult = "searchResult"
}