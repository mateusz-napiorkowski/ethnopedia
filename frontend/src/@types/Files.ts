export interface FileToDelete {
    originalFilename: string,
    newFilename: string,
    filePath: string,
    size: number,
    uploadedAt: Date,
    _id: string
}