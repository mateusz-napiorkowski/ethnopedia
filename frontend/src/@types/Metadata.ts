export interface Metadata {
    name: string;
    value: string;
    subcategories?: Metadata[];
    isSelectable?: boolean;
}