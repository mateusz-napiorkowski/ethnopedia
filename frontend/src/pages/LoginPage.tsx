import React, { useEffect, useState } from "react"
import { Formik, Form, Field, ErrorMessage } from "formik"
import * as Yup from "yup"
import { useNavigate } from "react-router-dom"
import { Toast, ToastToggle } from "flowbite-react"
import { HiExclamation } from "react-icons/hi"
import { useLoginMutation } from "../api/auth"
import { JWT, useUser } from "../providers/UserProvider"
import { jwtDecode } from "jwt-decode"

const LoginPage = () => {
    const loginMutation = useLoginMutation()
    const { setUserData } = useUser()

    const [showErrorToast, setShowErrorToast] = useState(false)
    const validationSchema = Yup.object().shape({
        username: Yup.string()
            .required("Nazwa użytkownika jest wymagana"),

        password: Yup.string()
            .required("Hasło jest wymagane"),
    })
    const navigate = useNavigate()
    const { mutate: loginUser } = loginMutation

    useEffect(() => {
        let timer: any
        if (showErrorToast) {
            timer = setTimeout(() => setShowErrorToast(false), 3000)
        }
        return () => clearTimeout(timer)
    }, [showErrorToast])


    return <section className="bg-gray-50 dark:bg-gray-900">
        {showErrorToast && (
            <div className={`fixed bottom-5 right-5 z-50 ${showErrorToast ? "animate-slide-in" : "animate-fade-out"}`}>
                <Toast>
                    <div
                        className="inline-flex h-8 w-8 items-center justify-center rounded-lg bg-red-100 text-red-500 dark:bg-red-700 dark:text-red-200">
                        <HiExclamation />
                    </div>
                    <div className="mx-3 text-sm font-normal">Nieprawidłowa nazwa użytkownika lub hasło.</div>
                    <ToastToggle onClick={() => setShowErrorToast(false)} />
                </Toast>
            </div>
        )}

        <div
            className="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0
                dark:border-gray-600">
            <div
                className="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800
                    dark:border-gray-600 border border-gray-200">
                <div className="p-6 space-y-4 md:space-y-6 sm:p-8">
                    <h1 className="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white">
                        Zaloguj się
                    </h1>
                    <Formik
                        initialValues={{ username: "", password: "" }}
                        validationSchema={validationSchema}
                        onSubmit={(values, { setSubmitting }) => {
                            const { username, password } = values

                            loginUser({ username, password }, {
                                onSuccess: (response) => {
                                    const { token } = response.data
                                    localStorage.setItem("token", token)
                                    const decodedToken = jwtDecode<JWT>(token)
                                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                                    navigate("/")
                                },
                                onError: (error: any) => {
                                    console.error(error)
                                    setShowErrorToast(true)
                                    setSubmitting(false)
                                },
                            })
                        }}
                    >
                        {({ errors, touched }) => (
                            <Form className="space-y-4 md:space-y-6">
                                <div>
                                    <label htmlFor="username"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Nazwa użytkownika
                                    </label>
                                    <Field name="username" id="username"
                                           className={`bg-gray-50 border ${errors.username && touched.username ? "border-red-500 bg-red-50" : "border-gray-300"}
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="Nazwa użytkownika" />
                                    <ErrorMessage name="username" component="div" className="text-red-600 text-sm" />
                                </div>
                                <div>
                                    <label htmlFor="password"
                                           className="block mb-2 text-sm font-medium text-gray-900 dark:text-white">
                                        Hasło
                                    </label>
                                    <Field type="password" name="password" id="password"
                                           className={`bg-gray-50 border ${errors.password && touched.password ? "border-red-500 bg-red-50" :
                                               "border-gray-300"}
                                                                                  text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600
                                                                                  focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700
                                                                                  dark:border-gray-600 dark:placeholder-gray-400 dark:text-white
                                                                                  dark:focus:ring-blue-500 dark:focus:border-blue-500`}
                                           placeholder="••••••••" />
                                    <ErrorMessage name="password" component="div"
                                                  className="text-red-600 text-sm" />
                                </div>
                                <button type="submit"
                                        className="w-full text-white bg-primary-600 hover:bg-sky-400 bg-sky-500">
                                    Zaloguj się
                                </button>
                            </Form>
                        )}
                    </Formik>
                    <p className="text-sm font-light text-gray-500 dark:text-gray-400">
                        Nie masz konta?{" "}
                        <span className="ml-1 font-medium text-primary-600 hover:underline dark:text-primary-500
                             cursor-pointer" onClick={() => navigate("/register")}>
                                Zarejestruj się.
                            </span>
                    </p>
                </div>
            </div>
        </div>
    </section>
}

export default LoginPage
