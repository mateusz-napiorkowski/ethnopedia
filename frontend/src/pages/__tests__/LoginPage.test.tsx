import { render, waitForElementToBeRemoved } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "react-query";
import LoginPage from "../../pages/LoginPage";
import userEvent from "@testing-library/user-event";
import { LoginValues } from "src/@types/Auth";

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom') as any,
    useNavigate: () => mockUseNavigate,
}));

const mockLoginUser = jest.fn()
jest.mock("../../api/auth", () => ({
    loginUser: ({username, password}: LoginValues) => mockLoginUser({username, password}),
}));

const queryClient = new QueryClient();
const user = userEvent.setup()

const jwtToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3Rvd3kiLCJmaXJzdE5hbWUiOiJ0ZXN0b3d5IiwidXN"
    + "lcklkIjoiNjZiNjUwNmZiYjY0ZGYxNjVlOGE5Y2U2IiwiaWF0IjoxNzI0MTg0MTE0LCJleHAiOjE3MjUxODQxMTR9.fzHPaXFMzQTVUf9IdZ0G6oeiaecc"
    + "N-rDSjRS3kApqlA"

const renderPage = () => {    
    return render(
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[`/login`]}>
                <Routes>
                    <Route path="/login" element={<LoginPage/>}/>
                </Routes>  
            </MemoryRouter>
        </QueryClientProvider>       
    );
};

describe("LoginPage tests", () => {
    beforeEach(() => {
        queryClient.clear();
        jest.clearAllMocks()
    });

    it("should render initial state", () => {
        const {container} = renderPage();
        expect(container).toMatchSnapshot()
    });

    it("should show appropriate error when username input field is empty after login button is clicked", async () => {
        const {getByPlaceholderText, getByRole, getByText, queryByText} = renderPage();
        const passwordInputField = getByPlaceholderText(/••••••••/i)
        const loginButton = getByRole("button", {name: /zaloguj się/i})

        await user.type(passwordInputField, "password")
        await user.click(loginButton)

        expect(getByText(/nazwa użytkownika jest wymagana/i)).toBeInTheDocument()
        expect(queryByText(/hasło jest wymagane/i)).not.toBeInTheDocument()
    });

    it("should show appropriate error when password input field is empty after login button is clicked", async () => {
        const {getByPlaceholderText, getByRole, getByText, queryByText} = renderPage();
        const usernameInputField = getByPlaceholderText(/nazwa użytkownika/i)
        const loginButton = getByRole("button", {name: /zaloguj się/i})

        await user.type(usernameInputField, "username")
        await user.click(loginButton)

        expect(getByText(/hasło jest wymagane/i)).toBeInTheDocument()
        expect(queryByText(/nazwa użytkownika jest wymagana/i)).not.toBeInTheDocument()
    });

    it("should call loginUser with correct parameters when both input fields are filled in and login button is clicked", async () => {
        mockLoginUser.mockReturnValue({token: jwtToken})
        const {getByPlaceholderText, getByRole} = renderPage();
        const usernameInputField = getByPlaceholderText(/nazwa użytkownika/i)
        const passwordInputField = getByPlaceholderText(/••••••••/i)
        const loginButton = getByRole("button", {name: /zaloguj się/i})

        await user.type(usernameInputField, "username")
        await user.type(passwordInputField, "password")
        await user.click(loginButton)

        expect(mockLoginUser).toHaveBeenCalledWith({username: "username", password: "password"})
    });

    it("should call useNavigate('/register') after register button is clicked", async () => {
        const { getByText } = renderPage();
        const registerButton = getByText(/zarejestruj się/i)
        
        await user.click(registerButton)

        expect(mockUseNavigate).toHaveBeenCalledWith("/register")
    });

    it("should show error toast when user with provided username and password doesn't exist, toast should disappear after a few seconds", async () => {
        mockLoginUser.mockImplementation(() => {throw Error("Mocked loginUser error")})          
        const { getByPlaceholderText, getByRole, getByText, queryByText } = renderPage()
        const usernameInputField = getByPlaceholderText(/nazwa użytkownika/i)
        const passwordInputField = getByPlaceholderText(/••••••••/i)
        const loginButton = getByRole("button", {name: /zaloguj się/i})

        await user.type(usernameInputField, "username")
        await user.type(passwordInputField, "password")
        await user.click(loginButton)

        expect(getByText(/nieprawidłowa nazwa użytkownika lub hasło./i)).toBeInTheDocument()
        await waitForElementToBeRemoved(() => queryByText(/nieprawidłowa nazwa użytkownika lub hasło./i), {timeout: 5000})
        expect(queryByText(/nieprawidłowa nazwa użytkownika lub hasło./i)).not.toBeInTheDocument()
    })

    it("should hide error toast when user clicks the 'X' button on it", async () => {
        mockLoginUser.mockImplementation(() => {throw Error("Mocked loginUser error")})          
        const { getByPlaceholderText, getByRole, getByLabelText, queryByText } = renderPage()
        const usernameInputField = getByPlaceholderText(/nazwa użytkownika/i)
        const passwordInputField = getByPlaceholderText(/••••••••/i)
        const loginButton = getByRole("button", {name: /zaloguj się/i})

        await user.type(usernameInputField, "username")
        await user.type(passwordInputField, "password")
        await user.click(loginButton)
        await user.click(getByLabelText("Close"))

        expect(queryByText(/nieprawidłowa nazwa użytkownika lub hasło./i)).not.toBeInTheDocument()
    })
});
