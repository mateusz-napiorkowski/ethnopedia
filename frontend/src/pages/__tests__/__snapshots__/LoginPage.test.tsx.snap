// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage tests should render initial state 1`] = `
<div>
  <section
    class="bg-gray-50 dark:bg-gray-900"
  >
    <div
      class="flex flex-col items-center justify-center px-6 py-8 mx-auto md:h-screen lg:py-0 dark:border-gray-600"
    >
      <div
        class="w-full bg-white rounded-lg shadow dark:border md:mt-0 sm:max-w-md xl:p-0 dark:bg-gray-800 dark:border-gray-600 border border-gray-200"
      >
        <div
          class="p-6 space-y-4 md:space-y-6 sm:p-8"
        >
          <h1
            class="text-xl font-bold leading-tight tracking-tight text-gray-900 md:text-2xl dark:text-white"
          >
            Zaloguj się
          </h1>
          <form
            action="#"
            class="space-y-4 md:space-y-6"
          >
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="username"
              >
                Nazwa użytkownika
              </label>
              <input
                class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="username"
                name="username"
                placeholder="Nazwa użytkownika"
                value=""
              />
            </div>
            <div>
              <label
                class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
                for="password"
              >
                Hasło
              </label>
              <input
                class="bg-gray-50 border border-gray-300 text-gray-900 sm:text-sm rounded-lg focus:ring-primary-600 focus:border-primary-600 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                id="password"
                name="password"
                placeholder="••••••••"
                type="password"
                value=""
              />
            </div>
            <button
              class="color-button w-full"
              data-testid="login-button"
              type="submit"
            >
              Zaloguj się
            </button>
          </form>
          <p
            class="text-sm font-light text-gray-500 dark:text-gray-400"
          >
            Nie masz konta?
             
            <span
              class="ml-1 font-medium text-primary-600 hover:underline dark:text-primary-500 cursor-pointer"
            >
              Zarejestruj się.
            </span>
          </p>
        </div>
      </div>
    </div>
  </section>
</div>
`;
